src/service.rs

```diff
@@ -623,109 +645,95 @@ where
         proto_id: ProtocolId,
         address: SocketAddr,
         ty: SessionType,
-        remote_public_key: &Option<PublicKey>,
+        remote_pubkey: &Option<PublicKey>,
         version: &str,
     ) {
         debug!("service session [{}] proto [{}] open", id, proto_id);
 
-        // Global proto handle processing flow
-        if let Some(handle) = self.proto_handles.get_mut(&proto_id) {
-            handle.connected(
-                &mut self.service_context,
-                id,
-                address,
-                ty,
-                &remote_public_key,
-                &version,
-            );
-        } else if let Some(mut handle) = self.get_proto_handle(false, proto_id) {
-            handle.init(&mut self.service_context);
-            handle.connected(
-                &mut self.service_context,
-                id,
-                address,
-                ty,
-                &remote_public_key,
-                &version,
-            );
-            self.proto_handles.insert(proto_id, handle);
+        let session_context = SessionContext {
+            id,
+            address,
+            ty,
+            remote_pubkey: remote_pubkey.clone(),
+            version: version.to_owned(),
+        };
+
+        // Service proto handle processing flow
+        if !self.service_proto_handles.contains_key(&proto_id) {
+            if let Some(ProtocolHandle::Service(mut handle)) =
+                self.get_proto_handle(false, proto_id)
+            {
+                debug!("init service [{}] level proto [{}] handle", id, proto_id);
+                handle.init(&mut self.service_context);
+                self.service_proto_handles.insert(proto_id, handle);
+            }
+        }
+        if let Some(handle) = self.service_proto_handles.get_mut(&proto_id) {
+            handle.connected(&mut self.service_context, &session_context);
         }
 
         // Session proto handle processing flow
         // Regardless of the existence of the session level handle,
         // you **must record** which protocols are opened for each session.
-        let session_level_handle = match self.get_proto_handle(true, proto_id) {
-            Some(mut handle) => {
-                debug!("init session [{}] level proto [{}] handle", id, proto_id);
-                handle.init(&mut self.service_context);
-                handle.connected(
-                    &mut self.service_context,
-                    id,
-                    address,
-                    ty,
-                    &remote_public_key,
-                    &version,
-                );
-                Some(handle)
-            }
-            None => None,
-        };
+        if let Some(ProtocolHandle::Session(mut handle)) = self.get_proto_handle(true, proto_id) {
```

Comment:
Look at the comments, here you must record all the protocols opened by each session, used to control the behavior when closing the session